examples/llm/src/mosaic_gpt.py

```diff
@@ -86,26 +87,74 @@ class FlashCausalAttention(nn.Module):
     def __init__(self, cfg: DictConfig, device: Optional[str] = None):
         super().__init__()
         try:
-            from flash_attn.flash_attention import FlashMHA  # type: ignore
+            from flash_attn.flash_attention import (  # type: ignore
+                FlashAttention, FlashMHA)
         except ImportError as e:
             raise e
 
-        self.mhsa = FlashMHA(
-            embed_dim=cfg.d_model,
-            num_heads=cfg.n_heads,
-            attention_dropout=cfg.attn_pdrop,
-            bias=True,
-            batch_first=True,
-            causal=True,
-            device=device,
-        )
-        self.mhsa.out_proj._is_residual = True
+        self.attn_qk_ln = cfg.attn_qk_ln
+        self.d_model = cfg.d_model
+        self.n_heads = cfg.n_heads
+
+        if self.attn_qk_ln:
+            self.W_qkv = nn.Linear(self.d_model,
+                                   3 * self.d_model,
+                                   bias=True,
+                                   device=device)
+            self.causal_attn = FlashAttention(attention_dropout=cfg.attn_pdrop,
+                                              device=device)
+            self.out_proj = nn.Linear(self.d_model,
+                                      self.d_model,
+                                      bias=True,
+                                      device=device)
+
+            self.out_proj._is_residual = True  # type: ignore
+
+            self.q_ln = nn.LayerNorm(self.d_model, device=device)
+            self.k_ln = nn.LayerNorm(self.d_model, device=device)
+        else:
+            self.mhsa = FlashMHA(
+                embed_dim=cfg.d_model,
+                num_heads=cfg.n_heads,
+                attention_dropout=cfg.attn_pdrop,
+                bias=True,
+                batch_first=True,
+                causal=True,
+                device=device,
+            )
+            self.mhsa.out_proj._is_residual = True
 
     def forward(self, x, key_padding_mask, attn_mask=None):
         assert attn_mask is None
-        return self.mhsa(x,
-                         key_padding_mask=key_padding_mask,
-                         need_weights=False)
+        if self.attn_qk_ln:
+            qkv = self.W_qkv(x)
+
+            # Applying layernorm to qk
+            dtype = qkv.dtype
+            q, k, v = qkv.split(self.d_model, dim=-1)
+            q = self.q_ln(q)
+            k = self.k_ln(k)
+            qkv = torch.cat([q, k, v], dim=-1)
+            qkv = qkv.to(dtype)
```

Comment:
Whats the purpose of the `dtype` stuff here? Shouldn't autocast make sure that the activations are always 16-bit as they are being passed around?

Reply:
I think layernorms upcast everything to fp32. As a result, we need to downcast back to the original datatype (in particular flash doesn't like fp32)

Reply:
Hmm but wouldn't that mean all our networks with autocast have FP32 activations running around? That doesn't sound right...